sentry_streams/sentry_streams/adapters/arroyo/translator.py

```diff
@@ -0,0 +1,296 @@
+from sentry_streams.pipeline.window import (
+    MeasurementUnit,
+    SlidingWindow,
+    TumblingWindow,
+    Window,
+)
+
+TPayload = TypeVar("TPayload")
+TResult = TypeVar("TResult")
+
+logger = logging.getLogger(__name__)
+
+
+class ArroyoAccumulator:
+
+    def __init__(
+        self,
+        acc: Callable[[], Accumulator[Any, Any]],
+    ) -> None:
+        self.acc = acc
+
+    def initial_value(self) -> Any:
+        # instantaiate the underlying accumulator every time
+        self.instance = self.acc()
+
+        # get the fresh initial value each time
+        return self.instance.get_value()
+
+    # types still tbd
+    def accumulator(self, result: Any, value: BaseValue[RoutedValue]) -> RoutedValue:
+        self.instance.add(value.payload.payload)
+
+        routed = RoutedValue(
+            route=value.payload.route,
+            payload=self.instance.get_value(),
+        )
+
+        return routed
+
+
+class KafkaAccumulator:
+    """
+    Does internal bookkeeping of offsets and timestamps,
+    as well as all the Accumulator functions.
+    """
+
+    def __init__(self, acc: Callable[[], Accumulator[Any, Any]]):
+        self.acc = acc()
+        self.offsets: MutableMapping[Partition, int] = {}
+        self.timestamp = time.time()
+
+    def add(self, value: Any) -> Self:
+
+        payload = value.payload.payload
+        offsets: MutableMapping[Partition, int] = value.committable
+
+        self.acc.add(payload)
+
+        self.offsets.update(offsets)
+
+        return self
+
+    def get_value(self) -> Any:
+
+        return self.acc.get_value()
+
+    def merge(self, other: Self) -> Self:
+
+        for partition in other.offsets:
+            if partition in self.offsets:
+                self.offsets[partition] = max(self.offsets[partition], other.offsets[partition])
+
+            else:
+                self.offsets[partition] = other.offsets[partition]
+
+        self.acc.merge(other.acc)
+
+        return self
+
+    def get_offsets(self) -> MutableMapping[Partition, int]:
+
+        return self.offsets
+
+    def clear(self) -> None:
+
+        assert hasattr(self.acc, "clear")
+        self.acc.clear()
+        self.offsets = {}
+
+        return
+
+
+class WindowedReduce(
+    ProcessingStrategy[Union[FilteredPayload, TPayload]], Generic[TPayload, TResult]
+):
+    def __init__(
+        self,
+        window_size: float,
+        window_slide: float,
+        acc: Callable[[], Accumulator[Any, Any]],
+        next_step: ProcessingStrategy[TResult],
+        route: Route,
+    ) -> None:
+
+        window_count = math.ceil(window_size / window_slide)
+
+        self.window_count = window_count
+        self.window_size = int(window_size)
+        self.window_slide = int(window_slide)
+
+        self.next_step = next_step
+        self.start_time = time.time()
+        self.route = route
+
+        # build a list of Accumulators
+        self.largest_val: int = int(2 * self.window_size - self.window_slide)
```

Comment:
What is largest_val supposed to represent? Is it the highest watermark among the watermarks of all your windows ?